src/Lachain.Core/Blockchain/Pool/TransactionHashTrackerByNonce.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Lachain.Logger;
+using Lachain.Proto;
+
+namespace Lachain.Core.Blockchain.Pool
+{
+    public class TransactionHashTrackerByNonce : ITransactionHashTrackerByNonce
+    {
+        private static readonly ILogger<TransactionHashTrackerByNonce> Logger 
+            = LoggerFactory.GetLoggerForClass<TransactionHashTrackerByNonce>();
+        private readonly ConcurrentDictionary<AddressWithNonce, UInt256> _hashPerAddressWithNonce
+            = new ConcurrentDictionary<AddressWithNonce, UInt256>();
+
+        public TransactionHashTrackerByNonce() { }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        public bool TryAdd(TransactionReceipt receipt)
+        {
+            var hash = receipt.Hash;
+            var nonce = receipt.Transaction.Nonce;
+            var from = receipt.Transaction.From;
+            var key = new AddressWithNonce(from, nonce);
+            if (_hashPerAddressWithNonce.ContainsKey(key)) return false;
```

Comment:
Is checking for the presence of the `key` required? I think, if `key` is already present `_hashPerAddressWithNonce.TryAdd(key, hash);` will return `false`?

Reply:
As the txes come from outside blockchain via api, it is safer to check. And here it is expected that the key is absent, because we are trying to add the key.